### DISCUSSION TITLE
iterate over returned array with urls

### DISCUSSION THREAD

**Discussion by Sembauke** [cypress-io/cypress#16284] [Questions and Help] [answered]
I have a task that returns an array of URLs.
This task returns the array in about 6 to 15 seconds.

Now I want to be able to iterate over these URLs.  

What I've tried: 

- I tried using a forEach loop outside the test like in  [#521](https://github.com/cypress-io/cypress-documentation/issues/521) - Having a `let challenges = []` and then setting the array value after the task yields won't work because a forEach loop runs immediately.

What I want it to do:

- I want it to test/iterate over every value in the array.
- I want to run multiple tests on the same url.

If I need to provide any additional info just ask.

```/* global cy */
describe('Visit APIs and Microservices', () => {
  before(() => {
    cy.task('getCurriculum', 'english').then(curriculum => {
      cy.task('scopeCurriculum', {
        curriculum,
        superblock: 'apis-and-microservices'
      }).as('challenges');
    });
  });

  it(`has access to the text`, () => {
    cy.get('@challenges').then(challenges => {
      describe(`visit challenge`, () => {
        challenges.forEach(challenge => {
          it(`should work correctly`, () => {
            cy.visit(challenge);
          });
        });
      });
    });
  });
});
```

**Comment by bahmutov** [comment, marked as answer]
Unfortunately as you have written, this won't work - as the spec will reload when the URL to visit changes, and having a test `it` inside another test `it` is going to create an error. 

Instead grab the URLs to visit in the plugins file, set as a config or env value and iterate to create tests. See the several "Dynamic..." recipes in our repo https://github.com/cypress-io/cypress-example-recipes#fundamentals

**Reply by Sembauke** [reply]
Thanks bahmutov.

**Comment by Sembauke** [comment]
So for the ones that are here to for the same issue this is what I did:

1. Instead of passing an array by the task in the spec file I made it write an object to a file in the fixture folder. You can do this with [filesystem from node](https://nodejs.org/api/fs.html#fs_filehandle_writefile_data_options). You do have to stringify your object into JSON.
2.  After that you can require the file in one of the spec files `const yourObject = require('path)`  

now you can loop over yourObejct with a forEach loop
